### PR TITLE
Add time to frame tracking to hot run

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Add FirstFrameTimer and wire it up around restart 'R' and hot reload 'r'.
- [x] Print the full time from key press to first frame after a restart / hot reload

This is the same number that we will report on benchmarks. Example output:

```
Hot reload time to frame: 0.69 seconds
```

```
Restart time to frame: 3.71 seconds
```

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@sethladd @devoncarew

File: packages/flutter_tools/lib/src/hot.dart [hunk 1]
```diff
@@ -29,6 +29,42 @@ String getDevFSLoaderScript() {
...
+    sw.reset();
+    sw.start();
+    _subscription = serviceProtocol.onExtensionEvent.listen(_onExtensionEvent);
+  }
+
+  /// Returns a Future which completes after the first frame event is received.
+  Future<Null> firstFrame() async => _completer.future;
+
+  void _onExtensionEvent(Event event) {
+    if (event.extensionKind == 'Flutter.FirstFrame') {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: nit: no braces

File: packages/flutter_tools/lib/src/hot.dart [hunk 2]
```diff
@@ -29,6 +29,42 @@ String getDevFSLoaderScript() {
...
+    _completer.complete(null);
+  }
+
+  Duration get elapsed {
+    assert(!sw.isRunning);
+    return sw.elapsed;
+  }
+
+  final Observatory serviceProtocol;
+  final Stopwatch sw = new Stopwatch();
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: `sw` ==> `stopwatch`?

File: packages/flutter_tools/lib/src/hot.dart [hunk 3]
```diff
@@ -347,16 +383,21 @@ class HotRunner extends ResidentRunner {
...
-      Status restartStatus = logger.startProgress('Restarting application...');
       await _launchFromDevFS(_package, _mainPath);
-      restartStatus.stop(showElapsedTime: true);
     }
+    Status restartStatus =
+        logger.startProgress('Waiting for application to start...');
+    // Wait for the first frame to be rendered.
+    await firstFrameTimer.firstFrame();
+    restartStatus.stop(showElapsedTime: true);
+    printStatus('Restart time to frame: '
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: `Restart time to frame` ==> `Restart time`?

File: packages/flutter_tools/lib/src/hot.dart [hunk 4]
```diff
@@ -410,6 +453,9 @@ class HotRunner extends ResidentRunner {
       return false;
     }
     reassembleStatus.stop(showElapsedTime: true);
+    await firstFrameTimer.firstFrame();
+    printStatus('Hot reload time to frame: '
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: `Hot reload time to frame` ==> `Hot reload time`?

File: packages/flutter_tools/lib/src/hot.dart [hunk 5]
```diff
@@ -29,6 +29,42 @@ String getDevFSLoaderScript() {
...
+  FirstFrameTimer(this.serviceProtocol);
+
+  void start() {
+    sw.reset();
+    sw.start();
+    _subscription = serviceProtocol.onExtensionEvent.listen(_onExtensionEvent);
+  }
+
+  /// Returns a Future which completes after the first frame event is received.
+  Future<Null> firstFrame() async => _completer.future;
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: the async is not necessary here

**Comment by devoncarew** [comment]
lgtm w/ some suggestions for verbiage

**Comment by eseidelGoogle** [comment]
Consider using ms to set expectations for these to be small. :)  It's also possible that seconds is the correct expectation.

**Comment by sethladd** [comment]
I like the suggestion to use ms. :)

File: packages/flutter_tools/lib/src/hot.dart (hunk 1 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/hot.dart (hunk 2 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/hot.dart (hunk 5 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/hot.dart (hunk 4 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

File: packages/flutter_tools/lib/src/hot.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/hot.dart]: Done.

**Pull request merged**